raiden-ts/src/transfers/mediate/types.ts

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { AddressZero, Zero } from '@ethersproject/constants';
+import * as t from 'io-ts';
+
+import type { Channel } from '../../channels';
+import type { UInt } from '../../utils/types';
+import { Address, decode, Int } from '../../utils/types';
+
+/**
+ * Defines the interface for a fee model, which receives and validates a given config,
+ * and calculates the fees for given config, channel state and input transfer amount.
+ * Generics in this type are important:
+ * - Config declares the configuration type expected for this specific feeModel component
+ * - Schedule declares the type for the output of [schedule] member function, which should be an
+ *    object to be merged with all other models and used as PFSFeeUpdate['fee_schedule'] payload
+ */
+export interface FeeModel<Config, Schedule> {
+  /** name of the fee model */
+  readonly name: string;
+  readonly emptySchedule: Schedule;
+  /**
+   * Validates and decodes the config; throws if expected config schema doesn't fit
+   *
+   * @param config - any object/value to validate as config
+   * @returns Decoded config, to be used in the other functions
+   */
+  decodeConfig: (config: unknown, defaultConfig?: unknown) => Config;
+  /**
+   * Given (validated) config and channel states, return a function to calculate this model's fees
+   * for each incoming transfer amount
+   *
+   * @param config - Validated and decoded config by [decodeConfig]
+   * @param channelIn - Channel state where transfer got received
+   * @param channelOut - Channel state where transfer is supposed to be forwarded through
+   * @returns Function which receives input amounts and returns fee
+   */
+  fee: (
+    config: Config,
+    channelIn: Channel,
+    channelOut: Channel,
+  ) => (amountIn: UInt<32>) => Int<32>;
+  /**
+   * Calculate this model's slice of the payload to PFSFeeUpdate messages
+   *
+   * @param config - Validated config
+   * @param channel - Channel state to calculate FeeUpdate for
+   * @returns A property of the fee_schedule member of [[PFSFeeUpdate]]
+   */
+  schedule: (config: Config, channel: Channel) => Schedule;
+}
+
+type ConfigOf<M extends FeeModel<any, any>> = ReturnType<M['decodeConfig']>;
+type ScheduleOf<M extends FeeModel<any, any>> = M['emptySchedule'];
+const mapCodec = t.union([t.undefined, t.record(t.string, t.record(t.string, t.unknown))]);
```

Comment:
This is specific to the standard fee calculator isn't? Could you try to push it down closer to the other one? It is not of relevance for the code that follows. Furthermore I would please you to rename it somehow as the name is somehow generic without much information about its actual purpose. I was only able to understand it because I know what it should be.

Reply:
Yeah, this came up when I moved the upper type helpers, I'll take it inside the function. Good catch.

Reply:
:heavy_check_mark: